pkg/permissions/permissions.go

```diff
@@ -1,28 +1,70 @@
 package permissions
 
 import (
+	"encoding/json"
 	"fmt"
 
 	"github.com/cozy/cozy-stack/pkg/consts"
 	"github.com/cozy/cozy-stack/pkg/couchdb"
 	"github.com/cozy/cozy-stack/pkg/couchdb/mango"
 	"github.com/cozy/cozy-stack/web/jsonapi"
+	"github.com/labstack/echo"
 )
 
 // Permission is a storable object containing a set of rules and
 // several codes
 type Permission struct {
-	PID           string            `json:"_id,omitempty"`
-	PRev          string            `json:"_rev,omitempty"`
-	ApplicationID string            `json:"application_id"`
-	Permissions   Set               `json:"permissions,omitempty"`
-	ExpiresAt     int               `json:"expires_at,omitempty"`
-	Codes         map[string]string `json:"codes,omitempty"`
+	PID         string            `json:"_id,omitempty"`
+	PRev        string            `json:"_rev,omitempty"`
+	Type        string            `json:"type,omitempty"`
+	SourceID    string            `json:"source_id,omitempty"`
+	Permissions *Set              `json:"permissions,omitempty"`
+	ExpiresAt   int               `json:"expires_at,omitempty"`
+	Codes       map[string]string `json:"codes,omitempty"`
 }
 
+const (
+	// TypeRegister is the value of Permission.Type for the temporary permissions
+	// allowed by registerToken
+	TypeRegister = "register"
+
+	// TypeApplication if the value of Permission.Type for an application
+	TypeApplication = "app"
+
+	// TypeSharing if the value of Permission.Type for a share permission doc
+	TypeSharing = "share"
+
+	// TypeOauth if the value of Permission.Type for a oauth permission doc
+	TypeOauth = "oauth"
+)
+
 // Index is the necessary index for this package
 // used in instance creation
-var Index = mango.IndexOnFields("application_id")
+var Index = mango.IndexOnFields("source_id", "type")
+
+const shareByCView = "byToken"
+
+// Views are athe necessary views for this package
+// used in instance creation
+var Views = couchdb.Views{
+	shareByCView: couchdb.View{
+		Map: `
+function(doc){
+	if(doc.type === "share" && doc.codes)
+		Object.keys(doc.codes).forEach(function(k){ emit(doc.codes[k]) })
+}`,
```

Comment:
Impossible to iterate over the keys with mango, we would need a sharing doc / recipient